services/web/server/src/simcore_service_webserver/rate_limiting.py

```diff
@@ -0,0 +1,66 @@
+import attr
+from functools import wraps
+from datetime import datetime, timedelta
+
+from aiohttp import web
+
+
+def global_rate_limit_route(reqs: int, interval_seconds: int):
+    """
+    Limits the requests per given interval to this endpoint
+    from all incoming sources.
+    Used to prevent abuse of unauthenticated endpoints.
+
+    reqs: number of max requests per total interval
+    interval_seconds: interval expressed in seconds
+    """
+
+    # compute the amount of requests per
+    def internal(decorated_function):
+        @attr.s(auto_attribs=True)
+        class Context:
+            max_allowed: int  # maimum allowed requests per interval
+            remaining: int  # remaining requests
+            rate_limit_reset: int  # utc timestamp
+
+        context = Context(max_allowed=reqs, remaining=reqs, rate_limit_reset=0)
+
+        @wraps(decorated_function)
+        async def wrapper(*args, **kwargs):
+            utc_now = datetime.utcnow()
+            current_utc_timestamp = datetime.timestamp(utc_now)
+
+            # reset counter & first time initialization
+            if current_utc_timestamp >= context.rate_limit_reset:
+                context.rate_limit_reset = datetime.timestamp(
+                    utc_now + timedelta(seconds=interval_seconds)
+                )
+                context.remaining = context.max_allowed
+
+            if (
+                current_utc_timestamp <= context.rate_limit_reset
+                and context.remaining <= 0
+            ):
+                # show error and return from here
+                return web.json_response(
```

Comment:
Instead of returning a response, I would raise an exception 
```python
raise HTTPForbidden(headers={ "X-Rate...": ... }, reason="API rate limit exceeded" )
```
This way a) it clearly states an error and b) the middleware will take care of enveloping and formatting the error correctly. See how the middleware treats this [here](https://github.com/ITISFoundation/osparc-simcore/blob/a957016dc8765b0b851bb67cb1f81ed7decf4ff8/packages/service-library/src/servicelib/rest_middlewares.py)

Reply:
migrated to an exception